### PR TITLE
Transforms in CapsDataset

### PR DESCRIPTION
**Pull request by thibaultdvx**
PR on `CapsDataset`, `Preprocessing` and `Transforms`.

For `CapsDataset`:

- Add `label` and `masks` arguments in `CapsDataset`: `label` can be either `None` (reconstruction), a mask (segmentation) or a column of the tsv `data` (classification, regression). `masks` are masks that are potentially necessary to compute some transforms. Consequently, I added methods `_get_masks`  and `_get_label` that are called in the `__getitem__`.
- Change `CapsDataset`'s `__getitem__` to handle TorchIO's `Subject` (following #686).
- As the `CapsDataset` class was growing pretty big, I took `check_preprocessing` and `prepare_data` methods out of it. `check_preprocessing` is now in `CapsReader`; and `prepare_data` is now a function that takes a `CapsDataset` as an input.
- Remove `CapsDatasetSample` as `CapsDataset`'s `__getitem__` now returns a `Sample` object (following #686).

For `Preprocessing`:

- Rename `BasePreprocessing` to `Preprocessing` and the enum `Preprocessing` to `PreprocessingMethod`. The motivation is to have `Preprocessing` in type hints.
- Remove mentions of preprocessings `t2-linear` and `t1-extensive` as they no longer exist.
- Remove arguments `use_uncropped_image` for `PreprocessingDTI` and `PreprocessingCustom` as it is only an argument for linear pipelines.

For `Transforms`:

- Remove `size_reduction`, `size_reduction_factor` and `normalize` in `Transforms`, as equivalent transforms can easily be passed.
- Add unittests to test the `Transforms` object.

Sorry for the huge PR. Besides, these are only suggestions, feel free to question them!

### CODE REVIEW

**Comment by thibaultdvx** [comment]
Ready for

File: clinicadl/data/datasets/caps_dataset.py [hunk 1]
```diff
@@ -50,24 +54,26 @@ class CapsDataset(Dataset):
...
     """
 
     def __init__(
         self,
-        caps_directory: Path,
-        preprocessing: BasePreprocessing,
-        transforms: Transforms,
-        data: Optional[Union[pd.DataFrame, Path]] = None,
+        caps_directory: Union[str, Path],
+        preprocessing: Preprocessing,
```
Comment (camillebrianceau) [inline clinicadl/data/datasets/caps_dataset.py]: maybe we can give a default preprocessing like `PreprocessingT1()` WDYT ?

File: clinicadl/data/datasets/caps_dataset.py [hunk 2]
```diff
@@ -76,41 +82,81 @@ def __init__(
...
+            'sub-001/ses-M000/sub-001_ses-M000_T1w.nii.gz' and `label="seg"`, it will look for the associated
+            mask in 'sub-001/ses-M000/sub-001_ses-M000_seg.nii.gz'.\n
+            If None, no label will be used (e.g. for reconstruction).
+        masks : Optional[list[str]] (optional, default=None)
+            Potential subject-specific masks that are useful to compute some transforms. CapsDataset
+            will look for masks with that values as a suffix. For example, if the image of the subject
+            'sub-001' for the session 'ses-M000' is in 'sub-001/ses-M000/sub-001_ses-M000_T1w.nii.gz'
+            and `masks=["brain", "hippocampus"]`, it will look for the masks in
+            'sub-001/ses-M000/sub-001_ses-M000_brain.nii.gz' and 'sub-001/ses-M000/sub-001_ses-M000_hippocampus.nii.gz'.
         """
```
Comment (camillebrianceau) [inline clinicadl/data/datasets/caps_dataset.py:115]:
```suggestion
caps_directory = Path(caps_directory) # maybe to put in the CapsReader
```

File: clinicadl/data/datasets/caps_dataset.py [hunk 3]
```diff
@@ -153,23 +199,23 @@ def from_json(cls, json_path: Path):
...
+            "samples_per_image": self._samples_per_image,
             "participants": self.df[PARTICIPANT_ID].nunique(),
             "sessions": self.df[SESSION_ID].nunique(),
             "preprocessing": self.preprocessing.model_dump(),
             "extraction": self.extraction.model_dump(),
         }
 
     def _get_df_from_input(
-        self, data: Optional[Union[pd.DataFrame, Path]] = None
+        self, data: Optional[Union[pd.DataFrame, Path]]
```
Comment (camillebrianceau) [inline clinicadl/data/datasets/caps_dataset.py]:
```suggestion
        self, data: Optional[Union[pd.DataFrame, Path, str]]
```

**Review by camillebrianceau — APPROVED** [review]
LGTM ! Thanks :)

File: clinicadl/data/datasets/caps_dataset.py (hunk 1 above)
Comment (thibaultdvx) [reply clinicadl/data/datasets/caps_dataset.py]: Yes, I think T1 is the most common (at least in our lab).

File: clinicadl/data/datasets/caps_dataset.py (hunk 2 above)
Comment (thibaultdvx) [reply clinicadl/data/datasets/caps_dataset.py:115]: It is done in the CapsReader yes: 

https://github.com/aramis-lab/clinicadl/blob/dc3443ee7ddf50aa82cec62c52ee9a9c9d075d09/clinicadl/data/readers/reader.py#L37

File: clinicadl/data/datasets/caps_dataset.py (hunk 1 above)
Comment (thibaultdvx) [reply clinicadl/data/datasets/caps_dataset.py]: Done

**Pull request merged**